### PR TITLE
WIP: client certificate revocation checking support.

### PR DESCRIPTION
**Pull request by cpu**
**Note: left as a draft since it relies on unreleased Webpki code, and because there are some open questions left to resolve**

This branch is an initial implementation of client certificate revocation checking using the recently added webpki CRL support.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,3 +7,6 @@ members = [
 ]
 exclude = ["admin/rustfmt"]
 resolver = "2"
+
+[patch.crates-io]
+webpki = { package='rustls-webpki', git = 'https://github.com/cpu/webpki.git', branch = 'cpu-crl-support-rustls-fixes' }
```
Comment (cpu) [inline Cargo.toml:12]: We can change this to use `main` from `rustls/webpki` once https://github.com/rustls/webpki/pull/92 lands.

File: examples/src/bin/tlsserver-mio.rs [hunk 2]
```diff
@@ -429,6 +429,8 @@ Options:
                         to certificate.  Optional.
     --auth CERTFILE     Enable client authentication, and accept certificates
                         signed by those roots provided in CERTFILE.
+    --crl CRLFILE ...   Perform client certificate revocation checking using the DER-encoded
+                        CRLFILE. May be used multiple times.
```
Comment (cpu) [inline examples/src/bin/tlsserver-mio.rs:433]: It probably makes sense to error if `--crl` is provided and not `--auth` - I haven't done that in this branch to keep things simple while we figure out some other more interesting bits.

File: rustls/src/error.rs [hunk 3]
```diff
@@ -66,6 +66,10 @@ pub enum Error {
     /// implementation.
     InvalidCertificate(CertificateError),
 
+    /// A provided certificate revocation list (CRL) was invalid.
+    // TODO(@cpu): provide more detail, maybe with a sub-enum, or by exposing webpki::Error?
+    InvalidCrl,
```
Comment (cpu) [inline rustls/src/error.rs:71]: This is one point I'd appreciate some guidance on - How do we want to handle the `webpki::Error`s returned by `CertRevocationList::from_der`.

We could:

* Follow the lead of `CertificateError` and make a sub-enum that has Rustls-specified versions of the possible error returns, and then impl a  `TryFrom<webpki::Error>` or something like `verify.rs`'s [`pki_error`](https://github.com/rustls/rustls/blob/5053fde25bb0fde5715a95601a62c8fae626600e/rustls/src/verify.rs#L653-L667) fn.
* Change `InvalidCrl` to something like `InvalidCrl(Arc<dyn StdError + Send + Sync>)` and wrap the `webpki::Error` in an arc as a generic error
* Leave it as-is and not provide more specificity.
* Something else?

File: rustls/tests/common/mod.rs [hunk 4]
```diff
@@ -284,7 +284,8 @@ pub fn get_client_root_store(kt: KeyType) -> RootCertStore {
 pub fn make_server_config_with_mandatory_client_auth(kt: KeyType) -> ServerConfig {
     let client_auth_roots = get_client_root_store(kt);
 
-    let client_auth = AllowAnyAuthenticatedClient::new(client_auth_roots);
+    // TODO(@cpu): allow building test configuration with CRLs.
+    let client_auth = AllowAnyAuthenticatedClient::new(client_auth_roots, Vec::default());
```
Comment (cpu) [inline rustls/tests/common/mod.rs:288]: I'll commit to writing some actual tests for this once we have the webpki side figured out and more of the questions in this branch are resolved.

File: rustls/src/verify.rs [hunk 5]
```diff
@@ -539,24 +540,28 @@ fn prepare<'a, 'b>(
 }
 
 /// A `ClientCertVerifier` that will ensure that every client provides a trusted
-/// certificate, without any name checking.
+/// certificate, without any name checking. Optionally, client certificates will
+/// have their revocation status checked using the DER encoded CRLs provided.
 pub struct AllowAnyAuthenticatedClient {
     roots: RootCertStore,
     subjects: Vec<DistinguishedName>,
+    crls: Vec<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs:548]: I'll probably put up a branch extending https://github.com/rustls/webpki/pull/92 with an owned representation for comparison purposes anyway. I think it will be required one way or the other for https://github.com/rustls/webpki/issues/80 and https://github.com/rustls/webpki/issues/81

File: rustls/src/verify.rs [hunk 6]
```diff
@@ -591,7 +596,7 @@ impl ClientCertVerifier for AllowAnyAuthenticatedClient {
             &webpki::TlsClientTrustAnchors(&trustroots),
             &chain,
             now,
-            &[], // TODO(@cpu): provide CRLs as appropriate.
+            &[], // TODO(@cpu): parse CRL DER and provide.
```
Comment (djc) [inline rustls/src/verify.rs]: I don't think we should keep this as a separate commit.

File: rustls/src/verify.rs [hunk 7]
```diff
@@ -15,6 +15,7 @@ use ring::digest::Digest;
 
 use std::sync::Arc;
 use std::time::SystemTime;
+use webpki::CertRevocationList;
```
Comment (djc) [inline rustls/src/verify.rs:18]: Style nit: should separate the webpki import from the std import block.

File: rustls/src/verify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/verify.rs:548]: I think even if we're going to keep the CRL data around as a `Vec<u8>` we should wrap that in a newtype wrapper (like `PrivateKey` and `Certificate`) to make usage somewhat more intentional.

File: rustls/src/verify.rs [hunk 8]
```diff
@@ -590,13 +590,24 @@ impl ClientCertVerifier for AllowAnyAuthenticatedClient {
         now: SystemTime,
     ) -> Result<ClientCertVerified, Error> {
         let (cert, chain, trustroots) = prepare(end_entity, intermediates, &self.roots)?;
+        let crls = self
+            .crls
+            .iter()
+            .map(|crl_der| webpki::CertRevocationList::from_der(crl_der))
+            .collect::<Vec<Result<CertRevocationList, webpki::Error>>>()
+            .into_iter()
```
Comment (djc) [inline rustls/src/verify.rs:598]: I think you can collect into `Result<Vec<_>, _>` directly (without the intermediate collection step).

File: rustls/src/verify.rs [hunk 9]
```diff
@@ -590,13 +590,24 @@ impl ClientCertVerifier for AllowAnyAuthenticatedClient {
...
+            .iter()
+            .collect::<Vec<&CertRevocationList>>();
         let now = webpki::Time::try_from(now).map_err(|_| Error::FailedToGetCurrentTime)?;
         cert.verify_is_valid_tls_client_cert(
             SUPPORTED_SIG_ALGS,
             &webpki::TlsClientTrustAnchors(&trustroots),
             &chain,
             now,
-            &[], // TODO(@cpu): parse CRL DER and provide.
+            crls.as_slice(),
```
Comment (djc) [inline rustls/src/verify.rs:610]: Maybe `verify_is_valid_tls_client_cert()` should take an `impl Iterator<Item = &CertRevocationList<'_>>` instead of a slice? Could avoid the intermediate allocations here.

File: rustls/src/error.rs (hunk 3 above)
Comment (djc) [reply rustls/src/error.rs:71]: To what extent are CRL errors from webpki disjoint from other webpki errors? I do feel like we would definitely want to separate "error while parsing CRL" from "client certificate was revoked according to CRL" in our rustls error messaging.

In terms of timing, maybe we should also try to parse the `CertRevocationList`s once at `AllowAuthenticatedClient` construction just to draw out any errors from parsing?

File: rustls/src/verify.rs (hunk 9 above)
Comment (ctz) [reply rustls/src/verify.rs:610]: perhaps something stronger than just `impl Iterator` would be needed, as the CRLs need to be iterated a bunch of times per `verify_is_valid_tls_client_cert`?

Comment (djc) [reply rustls/src/verify.rs:610]: Ahh yes, we'll need to loop over it once per certificate considered for chain building. I suppose in that case a slice would be best after all.

File: rustls/src/verify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs:548]: Good idea.

File: rustls/src/verify.rs (hunk 9 above)
Comment (cpu) [reply rustls/src/verify.rs:610]: For my own education what do you mean by "something stronger"? Is the appeal of maintaining the slice related to some overhead of an `impl Iterator` approach? I have no strong pref. here, just want to understand the trade-offs better.

Comment (ctz) [reply rustls/src/verify.rs:610]: I don't know how to write it precisely, but it'd be something like `impl Iterator<...> + Clone` (ie, it's an Iterator, but you can also clone the iterator to get more) or a function that returns a fresh `impl Iterator` each time.

File: rustls/src/error.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/error.rs:71]: I need to dig in to where the "client certificate was revoked" error is getting bubbled up through Rustls. I have a feeling it's being converted to a Rustls internal error already. The error I added/flagged for discussion here is strictly from the "error while parsing CRL" codepath.

For the parsing errors, I think we can emit:
* UnsupportedCrlVersion
* UnsupportedDeltaCrl
* UnsupportedIndirectCrl
* UnsupportedRevocationReason
* InvalidCrlNumber
* InvalidSerialNumber
* SignatureAlgorithmMismatch
* UnknownIssuer
* MalformedExtensions
* BadDerTime
* BadDer (like almost all parts of webpki :sweat_smile:)

Of these I think probably the "Unsupported" ones are the most interesting to a user and most disjoint from existing errors. Most of the others (minus InvalidCrlNuber) are already used by webpki elsewhere.

On the path of checking a cert for revocation with a CRL since we front-load most of the error checking for the CRL itself I think we only emit:

* CertRevoked
* IssuerNotCrlSigner
* InvalidSignatureForPublicKey

Of these only CertRevoked and IssuerNotCrlSigner are new errors not previously possible for Rustls code to see.

Comment (cpu) [reply rustls/src/error.rs:71]: > In terms of timing, maybe we should also try to parse the CertRevocationLists once at AllowAuthenticatedClient construction just to draw out any errors from parsing?

That's a good idea :+1:

File: examples/src/bin/tlsserver-mio.rs (hunk 2 above)
Comment (cpu) [reply examples/src/bin/tlsserver-mio.rs:434]: I added this quickly since it was a small change ultimately.

File: rustls/src/error.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/error.rs:71]: Implemented up-front parsing in my recent push.

File: rustls/src/verify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs]: Added the newtype wrapper in my recent push.

File: rustls/src/verify.rs (hunk 6 above)
Comment (cpu) [reply rustls/src/verify.rs]: Collapsed into one commit.

File: rustls/src/verify.rs (hunk 7 above)
Comment (cpu) [reply rustls/src/verify.rs]: This fell away when I introduced the local newtype to wrap `Vec<u8>` the ~two places we want to use the webpki type I changed to be explicit references to `webpki::CertRevocationList`.

File: rustls/src/verify.rs (hunk 8 above)
Comment (cpu) [reply rustls/src/verify.rs]: Good call, thanks. Fixed up.

File: rustls/src/verify.rs (hunk 9 above)
Comment (cpu) [reply rustls/src/verify.rs:606]: I haven't tackled this feedback yet (want to try some other stuff first) but it's on my radar.

File: rustls/src/verify.rs [hunk 10]
```diff
@@ -569,12 +589,21 @@ impl ClientCertVerifier for AllowAnyAuthenticatedClient {
         now: SystemTime,
     ) -> Result<ClientCertVerified, Error> {
         let (cert, chain, trustroots) = prepare(end_entity, intermediates, &self.roots)?;
+        let crls = self
+            .crls
+            .iter()
+            .map(|crl| crl.parse())
+            .collect::<Result<Vec<_>, _>>()?;
```
Comment (cpu) [inline rustls/src/verify.rs:596]: leaving a note we could `unwrap()` here now that we're pre-validating the CRL DER in `new`.

File: rustls/src/error.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/error.rs:71]: Started on benchmarking: https://github.com/rustls/webpki/pull/94

File: rustls/src/verify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs]: > The intuition here is that going from DER to the CRL view over the DER is cheap (especially since signature validation isn't done until a cert with an issuer matching the CRL is consulted for a serial), but that should be verified with benchmarking (I will pursue that separately).

Started on benchmarking: https://github.com/rustls/webpki/pull/94

Initial results show parsing the CRL and finding serials in the CRL are roughly equivalent in runtime, with both seeming fairly quick from my perspective.

File: rustls/src/error.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/error.rs:71]: > I need to dig in to where the "client certificate was revoked" error is getting bubbled up through Rustls. I have a feeling it's being converted to/masked by a Rustls internal error already.

Traced this out. We were turning the `webpki::Error::CertRevoked` return from the verifier into a `InvalidCertificate(CertificateError::Other(...))` err. I had to add a match arm in `verify.rs`'s `pki_error` for `webpki::RevokedCert`. Conveniently there was already a `CertificateError::Revoked` on the Rustls side to map it to. I force pushed an update to this branch to include that change and in the "happy" revocation path we now get a sensible error bubbled up to the user:

Server side:
```
cargo run --package rustls-examples --bin tlsserver-mio -- 
  -p 4443 
  --certs ./pki/localhost/cert.pem 
  --key ./pki/localhost/key.pem 
  --auth ./pki/minica.pem
  --crl ./pki/demo.crl
  --crl ./pki/demo.2.crl
  http
```

Client side:
```
cargo run --bin tlsclient-mio -- 
  -p 4443 
  --http localhost 
  --cafile ./pki/minica.pem 
  --auth-key ./pki/localhost/key.pem 
  --auth-certs ./pki/localhost/cert.pem
```

```
TLS error: AlertReceived(CertificateRevoked)
Connection closed
```

File: rustls/src/verify.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/verify.rs]: I guess the next question is, how does this stuff compare in performance to the other things a webpki client certificate verification has to run through? That is, what is the impact on this relative to the whole operation?

[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs]: > I guess the next question is, how does this stuff compare in performance to the other things a webpki client certificate verification has to run through?

Probably the most substantial comparison point would be to path building in general, though my expectation there is that signature validation will be the dominant factor. I'll give this some thought.

> Initial results show parsing the CRL and finding serials in the CRL are roughly equivalent in runtime, with both seeming fairly quick from my perspective.

Just noting that this has an obvious explanation I overlooked :facepalm: - When we parse CRLs we're also walking the entire set of revoked certificates to detect errors early.

Before I start on a more time intensive side quest to benchmark path building I'm going to rework the code under benchmark to avoid doing that up-front check so we can have a better sense of the speed of just parsing the base CRL vs parsing/searching the revoked certificates.

[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/verify.rs]: >  I'm going to rework the code under benchmark to avoid doing that up-front check so we can have a better sense of the speed of just parsing the base CRL vs parsing/searching the revoked certificates.

Pushed that change to https://github.com/rustls/webpki/pull/94 and the results now clearly show that parsing the base CRL is very fast relative to parsing and searching its revoked cert contents.
